apps/anoma_node/lib/node/transaction/ordering.ex

```diff
@@ -357,20 +579,483 @@ defmodule Anoma.Node.Transaction.Ordering do
+
+    # Recalculate watermarks for the keys involved
+    state_with_watermarks =
+      recalculate_and_update_watermarks_for_keys(
+        keys_to_update,
+        state_with_acquired_tx
+      )
+
+    try_finalize_transaction(tx_id, state_with_watermarks)
+  end
+
+  @spec handle_reservation_failure(
+          binary(),
+          integer(),
+          %{binary() => pid()},
+          [binary()],
+          t()
+        ) :: t()
+  defp handle_reservation_failure(
+         tx_id,
+         height,
+         failed_shard_pids,
+         keys_to_update,
+         state
+       ) do
+    # Reservation failed, rollback successful ones on shards
+    # failed_shard_pids contains pids up to the point of failure
+    release_reservations(failed_shard_pids, height)
+
+    # Remove the failed transaction entry from the state it was added to
+    state_without_failed_tx = %{
+      state
+      | tx_reservations: Map.delete(state.tx_reservations, tx_id)
+    }
+
+    # Recalculate watermarks - failure might unblock something (use argument)
+    state_with_watermarks =
+      recalculate_and_update_watermarks_for_keys(
+        keys_to_update,
+        state_without_failed_tx
+      )
+
+    try_finalize_transaction(tx_id, state_with_watermarks)
+  end
+
+  @spec block_reservations(
+          String.t(),
+          binary(),
+          [{:read | :write, binary()}]
+        ) :: :ok
+  defp block_reservations(node_id, tx_id, reservations) do
+    receive do
+      %EventBroker.Event{
+        body: %Node.Event{body: %__MODULE__.OrderEvent{tx_id: ^tx_id}}
+      } ->
+        request_reservations(node_id, tx_id, reservations)
+
+      _ ->
+        IO.puts("this should be unreachable")
+    end
+
+    EventBroker.unsubscribe_me([
+      Node.Event.node_filter(node_id),
+      this_module_filter(),
+      tx_id_filter(tx_id)
+    ])
+  end
+
+  # Helper to release reservations on Shards
+  @spec release_reservations(%{binary() => pid()}, integer()) :: :ok
+  defp release_reservations(shard_pids, height) do
+    Enum.each(shard_pids, fn {_key, shard_pid} ->
+      # Cast should be okay for unreserve as it's best-effort rollback
+      Shard.unreserve(shard_pid, height)
+    end)
+  end
+
+  # Recalculates watermarks for a list of keys and sends updates to shards if they advance.
+  @spec recalculate_and_update_watermarks_for_keys([binary()], t()) :: t()
+  defp recalculate_and_update_watermarks_for_keys(keys, state) do
+    Enum.reduce(keys |> Enum.uniq(), state, fn key, acc_state ->
+      recalculate_and_update_watermarks_for_key(key, acc_state)
+    end)
+  end
+
+  # Recalculates watermarks for a single key and sends updates to shards if they advance.
+  @spec recalculate_and_update_watermarks_for_key(binary(), t()) :: t()
+  defp recalculate_and_update_watermarks_for_key(key, state) do
+    current_wms =
+      Map.get(state.watermark_state, key, %{
+        read_watermark: -1,
+        write_watermark: -1
+      })
```

Comment:
The default read_watermark and write_watermark leaks everywhere this should be properly abstracted and not leak as what if we add other watermarks? Do we have to chase all of these?